File: src/main/kotlin/net/echo/little_johns/block/custom/TableSawBlock.java
package net.echo.little_johns.block.custom;


import com.mojang.serialization.MapCodec;
import net.echo.little_johns.stat.ModStats;
import net.minecraft.block.*;
import net.minecraft.entity.ai.pathing.NavigationType;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.fluid.FluidState;
import net.minecraft.fluid.Fluids;
import net.minecraft.item.ItemPlacementContext;
import net.minecraft.screen.NamedScreenHandlerFactory;
import net.minecraft.screen.ScreenHandlerContext;
import net.minecraft.screen.SimpleNamedScreenHandlerFactory;
import net.minecraft.screen.StonecutterScreenHandler;
import net.minecraft.state.StateManager;
import net.minecraft.state.property.BooleanProperty;
import net.minecraft.state.property.EnumProperty;
import net.minecraft.state.property.Properties;
import net.minecraft.text.Text;
import net.minecraft.util.ActionResult;
import net.minecraft.util.BlockMirror;
import net.minecraft.util.BlockRotation;
import net.minecraft.util.hit.BlockHitResult;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.util.math.random.Random;
import net.minecraft.util.shape.VoxelShape;
import net.minecraft.util.shape.VoxelShapes;
import net.minecraft.world.BlockView;
import net.minecraft.world.World;
import net.minecraft.world.WorldView;
import net.minecraft.world.tick.ScheduledTickView;
import org.jetbrains.annotations.Nullable;

public class TableSawBlock
        extends Block
        implements Waterloggable {
    public static final MapCodec<TableSawBlock> CODEC = TableSawBlock.createCodec(TableSawBlock::new);
    private static final BooleanProperty WATERLOGGED = Properties.WATERLOGGED;
    private static final Text TITLE = Text.translatable("container.little_johns.table_saw");
    public static final EnumProperty<Direction> FACING = HorizontalFacingBlock.FACING;
    private static final VoxelShape CENTER = Block.createCuboidShape(0.0, 14.0, 0.0, 16.0, 16.0, 16.0);
    private static final VoxelShape LEG_1 = Block.createCuboidShape(0.0, 0.0, 0.0, 2.0, 16.0, 2.0);
    private static final VoxelShape LEG_3 = Block.createCuboidShape(14.0, 0.0, 0.0, 16.0, 16.0, 2.0);
    private static final VoxelShape LEG_2 = Block.createCuboidShape(0.0, 0.0, 14.0, 2.0, 16.0, 16.0);
    private static final VoxelShape LEG_4 = Block.createCuboidShape(14.0, 0.0, 14.0, 16.0, 16.0, 16.0);
    private static final VoxelShape SHAPE = VoxelShapes.union(CENTER, LEG_1, LEG_2, LEG_3, LEG_4);

    @Override
    public MapCodec<TableSawBlock> getCodec() {
        return CODEC;
    }


    public TableSawBlock(Settings settings) {
        super(settings);
        this.setDefaultState((this.stateManager.getDefaultState()).with(FACING, Direction.NORTH).with(WATERLOGGED, false));
    }

    @Override
    public BlockState getPlacementState(ItemPlacementContext ctx) {
        return (BlockState)this.getDefaultState().with(FACING, ctx.getHorizontalPlayerFacing().getOpposite());
    }


    @Override
    protected ActionResult onUse(BlockState state, World world, BlockPos pos, PlayerEntity player, BlockHitResult hit) {
        if (!world.isClient) {
            player.openHandledScreen(state.createScreenHandlerFactory(world, pos));
            player.incrementStat(ModStats.INTERACT_WITH_TABLE_SAW);
        }

        return ActionResult.SUCCESS;
    }

    @Override
    @Nullable
    protected NamedScreenHandlerFactory createScreenHandlerFactory(BlockState state, World world, BlockPos pos) {
        return new SimpleNamedScreenHandlerFactory((syncId, playerInventory, player)
                -> new StonecutterScreenHandler(syncId, playerInventory, ScreenHandlerContext.create(world, pos)), TITLE);
    }


    @Override
    public VoxelShape getOutlineShape(BlockState state, BlockView world, BlockPos pos, ShapeContext context) {
        return SHAPE;
    }

    @Override
    protected boolean hasSidedTransparency(BlockState state) {
        return true;
    }

    @Override
    protected BlockRenderType getRenderType(BlockState state) {
        return BlockRenderType.MODEL;
    }

    @Override
    protected BlockState rotate(BlockState state, BlockRotation rotation) {
        return (BlockState)state.with(FACING, rotation.rotate(state.get(FACING)));
    }

    @Override
    protected BlockState mirror(BlockState state, BlockMirror mirror) {
        return state.rotate(mirror.getRotation(state.get(FACING)));
    }

    @Override
    protected void appendProperties(StateManager.Builder<Block, BlockState> builder) {
        builder.add(FACING);
        builder.add(WATERLOGGED);
    }

    protected BlockState getStateForNeighborUpdate(BlockState state, WorldView world, ScheduledTickView tickView, BlockPos pos, Direction direction, BlockPos neighborPos, BlockState neighborState, Random random) {
        if ((Boolean)state.get(WATERLOGGED)) {
            tickView.scheduleFluidTick(pos, Fluids.WATER, Fluids.WATER.getTickRate(world));
        }

        return super.getStateForNeighborUpdate(state, world, tickView, pos, direction, neighborPos, neighborState, random);
    }


    @Override
    protected FluidState getFluidState(BlockState state) {
        if (state.get(WATERLOGGED).booleanValue()) {
            return Fluids.WATER.getStill(false);
        }
        return super.getFluidState(state);
    }

    @Override
    protected boolean canPathfindThrough(BlockState state, NavigationType type) {
        return false;
    }


}

File: src/main/kotlin/net/echo/little_johns/block/custom/TableSawBlock.kt
package net.echo.little_johns.block.custom

import com.mojang.serialization.MapCodec
import net.echo.little_johns.screen.TableSawScreenHandler
import net.echo.little_johns.stat.ModStats
import net.minecraft.block.*
import net.minecraft.entity.ai.pathing.NavigationType
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.fluid.FluidState
import net.minecraft.fluid.Fluids
import net.minecraft.item.ItemPlacementContext
import net.minecraft.screen.*
import net.minecraft.state.StateManager
import net.minecraft.state.property.BooleanProperty
import net.minecraft.state.property.EnumProperty
import net.minecraft.state.property.Properties
import net.minecraft.text.Text
import net.minecraft.util.*
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.*
import net.minecraft.util.math.random.Random
import net.minecraft.util.shape.VoxelShape
import net.minecraft.util.shape.VoxelShapes
import net.minecraft.world.*
import net.minecraft.world.tick.ScheduledTickView
import org.jetbrains.annotations.Nullable

class TableSawBlock(settings: Settings) : Block(settings), Waterloggable {
    companion object {
        val CODEC: MapCodec<TableSawBlock> = createCodec(::TableSawBlock)
        val WATERLOGGED: BooleanProperty = Properties.WATERLOGGED
        val TITLE: Text = Text.translatable("container.little_johns.table_saw")
        val FACING: EnumProperty<Direction> = HorizontalFacingBlock.FACING

        private val CENTER: VoxelShape = Block.createCuboidShape(0.0, 14.0, 0.0, 16.0, 16.0, 16.0)
        private val LEG_1: VoxelShape = Block.createCuboidShape(0.0, 0.0, 0.0, 2.0, 16.0, 2.0)
        private val LEG_2: VoxelShape = Block.createCuboidShape(0.0, 0.0, 14.0, 2.0, 16.0, 16.0)
        private val LEG_3: VoxelShape = Block.createCuboidShape(14.0, 0.0, 0.0, 16.0, 16.0, 2.0)
        private val LEG_4: VoxelShape = Block.createCuboidShape(14.0, 0.0, 14.0, 16.0, 16.0, 16.0)
        private val SHAPE: VoxelShape = VoxelShapes.union(CENTER, LEG_1, LEG_2, LEG_3, LEG_4)
    }

    override fun getCodec(): MapCodec<TableSawBlock> = CODEC

    init {
        defaultState = stateManager.defaultState
            .with(FACING, Direction.NORTH)
            .with(WATERLOGGED, false)
    }

    override fun getPlacementState(ctx: ItemPlacementContext): BlockState {
        return defaultState.with(FACING, ctx.horizontalPlayerFacing.opposite)
    }

    override fun onUse(
        state: BlockState,
        world: World,
        pos: BlockPos,
        player: PlayerEntity,
        hit: BlockHitResult
    ): ActionResult {
        if (!world.isClient) {
            player.openHandledScreen(createScreenHandlerFactory(state, world, pos))
            player.incrementStat(ModStats.INTERACT_WITH_TABLE_SAW)
        }
        return ActionResult.SUCCESS
    }

    @Nullable
    override fun createScreenHandlerFactory(
        state: BlockState,
        world: World,
        pos: BlockPos
    ): NamedScreenHandlerFactory {
        return SimpleNamedScreenHandlerFactory(
            { syncId, inventory, player ->
                TableSawScreenHandler(syncId, inventory, ScreenHandlerContext.create(world, pos))
            },
            TITLE
        )
    }

    override fun getOutlineShape(
        state: BlockState,
        world: BlockView,
        pos: BlockPos,
        context: ShapeContext
    ): VoxelShape = SHAPE

    override fun hasSidedTransparency(state: BlockState): Boolean = true

    override fun getRenderType(state: BlockState): BlockRenderType = BlockRenderType.MODEL

    override fun rotate(state: BlockState, rotation: BlockRotation): BlockState {
        return state.with(FACING, rotation.rotate(state[FACING]))
    }

    override fun mirror(state: BlockState, mirror: BlockMirror): BlockState {
        return state.rotate(mirror.getRotation(state[FACING]))
    }

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
        builder.add(FACING, WATERLOGGED)
    }

    override fun getStateForNeighborUpdate(
        state: BlockState,
        world: WorldView,
        tickView: ScheduledTickView,
        pos: BlockPos,
        direction: Direction,
        neighborPos: BlockPos,
        neighborState: BlockState,
        random: Random
    ): BlockState {
        if (state[WATERLOGGED]) {
            tickView.scheduleFluidTick(pos, Fluids.WATER, Fluids.WATER.getTickRate(world))
        }
        return super.getStateForNeighborUpdate(state, world, tickView, pos, direction, neighborPos, neighborState, random)
    }

    override fun getFluidState(state: BlockState): FluidState {
        return if (state[WATERLOGGED]) {
            Fluids.WATER.getStill(false)
        } else {
            super.getFluidState(state)
        }
    }

    override fun canPathfindThrough(state: BlockState, type: NavigationType): Boolean = false
}
